rpcclient/roundrobin.go

```diff
@@ -0,0 +1,347 @@
+				log.Infof("Balancer: Established connection to RPC server %s",
+					hostAddress.Host)
+				break
+			}
+			close(rrb.waitToConnect)
+		}
+		wsConn, _ = rrb.GetWsConnection(hostAddress.Host)
+	} else {
+		rrb.NotifyConnStateChange(hostAddress, Ready, true)
+	}
+	log.Infof("Balancer: Connection pick for RPC %s",
+		hostAddress.Host)
+	return wsConn, hostAddress, err
+}
+
+//NotifyConnStateChange updates connState map for the given address.
+//Also updates the isReady field to indicate that at least one connection
+//with Ready state exists
+func (rrb *RoundRobinBalancer) NotifyConnStateChange(hostAdd *HostAddress, state ConnectionState, sync bool) {
+	if sync {
+		rrb.mu.Lock()
+	}
+	if state == Shutdown && !rrb.connConfig.HTTPPostMode {
+		delete(rrb.wsConns, hostAdd.Host)
+	}
+	if state == Ready {
+		rrb.isReady = true
+	} else {
+		rrb.isReady = false
+		for _, state := range rrb.connState {
+			if state == Ready {
+				rrb.isReady = true
+				break
+			}
+		}
+	}
+	rrb.connState[hostAdd.Host] = state
+	if sync {
+		rrb.mu.Unlock()
+	}
+}
+
+// Close closes all the socket connections in the wsConns list
+// if no host string passed.
+// This must be called during shutdown with empty string
+func (rrb *RoundRobinBalancer) Close(host string) {
+	rrb.mu.Lock()
+	if host != "" {
+		log.Tracef("Balancer: Disconnecting RPC client %s", host)
+		rrb.wsConns[host].Close()
+		rrb.NotifyConnStateChange(rrb.hostAddMap[host], Disconnected, false)
+		rrb.mu.Unlock()
+		return
+	}
+
+	for host := range rrb.wsConns {
+		if rrb.connState[host] != Shutdown {
+			log.Tracef("Balancer: Disconnecting RPC client %s", host)
+			rrb.NotifyConnStateChange(rrb.hostAddMap[host], Disconnected, false)
+			rrb.wsConns[host].Close()
+		}
+	}
+	rrb.mu.Unlock()
+}
+
+// NotifyReconnect will update the map for wsConns and update the connection state
+// for this address
+func (rrb *RoundRobinBalancer) NotifyReconnect(wsConn *websocket.Conn, hostAdd *HostAddress) {
+	rrb.mu.Lock()
+	rrb.connState[hostAdd.Host] = Ready
+	rrb.wsConns[hostAdd.Host] = wsConn
+	rrb.hostAddMap[hostAdd.Host].retryCount = 0
+	rrb.mu.Unlock()
+}
+
+// UpdateReconnectAttempt will increase retryattempt + 1
+// for this address
+func (rrb *RoundRobinBalancer) UpdateReconnectAttempt(hostAdd *HostAddress) {
+	rrb.mu.Lock()
+	rrb.hostAddMap[hostAdd.Host].retryCount++
+	rrb.mu.Unlock()
+}
+
+// IsAllDisconnected would return true if all hostaddress are marked with
+// connection state as Disconnected
+func (rrb *RoundRobinBalancer) IsAllDisconnected() bool {
+	rrb.mu.Lock()
+	for _, state := range rrb.connState {
+		if state != Disconnected && state != Shutdown {
+			rrb.mu.Unlock()
+			return false
+		}
+	}
+	rrb.mu.Unlock()
+	return true
+}
+
+//GetNextDisconnectedWsConn will iterate over the connection state map and
+//return the first wsConnection that has its state as Disconnected
+func (rrb *RoundRobinBalancer) GetNextDisconnectedWsConn() *HostAddress {
```

Comment:
The names of functions that are designed as getters should not be named with `Get` per [Effective Go Guidelines](https://golang.org/doc/effective_go.html#Getters) which is part of the [Code Contribution Guidelines](https://github.com/decred/dcrd/blob/master/docs/code_contribution_guidelines.md#ReqReading).

It should be `NextDisconnectedWsConn`.